<a href="https://colab.research.google.com/github/nicolafricker/TennisMatchAiPredictor/blob/main/TennisPrecition.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

url = 'https://raw.githubusercontent.com/JeffSackmann/tennis_atp/refs/heads/master/atp_players.csv'
player_df = pd.read_csv(url, dtype={'wikidata_id': str})
display(player_df.head())

player_id name_first name_last hand         dob  ioc  height wikidata_id
0     100001    Gardnar    Mulloy    R  19131122.0  USA   185.0      Q54544
1     100002     Pancho    Segura    R  19210620.0  ECU   168.0      Q54581
2     100003      Frank   Sedgman    R  19271002.0  AUS   180.0     Q962049
3     100004   Giuseppe     Merlo    R  19271011.0  ITA     NaN    Q1258752
4     100005    Richard  Gonzalez    R  19280509.0  USA   188.0      Q53554

# Task
Import the playerbase from "https://raw.githubusercontent.com/JeffSackmann/tennis_atp/refs/heads/master/atp_players.csv" and the rankings from "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_00s.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_10s.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_20s.csv", "https://raw.githubusercontent.com/JeffSackmann/tennis_atp/refs/heads/master/atp_rankings_70s.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_80s.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_90s.csv", and "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_current.csv".

## Define urls

### Subtask:
Create a list of all the URLs for the ranking data.


**Reasoning**:
Create a Python list containing the specified URLs for the ranking data.



In [ ]:
ranking_urls = [
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_00s.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_10s.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_20s.csv",
    "https://raw.githubusercontent.com/JeffSackmann/tennis_atp/refs/heads/master/atp_rankings_70s.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_80s.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_90s.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_current.csv"
]

## Load and concatenate data

### Subtask:
Iterate through the list of URLs, read each CSV file into a pandas DataFrame, and concatenate all the DataFrames into a single one.


**Reasoning**:
Iterate through the list of URLs, read each CSV file into a pandas DataFrame, and concatenate all the DataFrames into a single one.



In [ ]:
ranking_dfs = []
for url in ranking_urls:
    ranking_dfs.append(pd.read_csv(url))

ranking_df = pd.concat(ranking_dfs, ignore_index=True)
display(ranking_df.head())

ranking_date  rank  player  points
0      20000110     1  101736  4135.0
1      20000110     2  102338  2915.0
2      20000110     3  101948  2419.0
3      20000110     4  103017  2184.0
4      20000110     5  102856  2169.0

## Display data

### Subtask:
Display the head of the combined DataFrame to verify the data is loaded correctly.


**Reasoning**:
Display the head of the combined DataFrame to verify the data is loaded correctly.



In [ ]:
display(ranking_df.head())

ranking_date  rank  player  points
0      20000110     1  101736  4135.0
1      20000110     2  102338  2915.0
2      20000110     3  101948  2419.0
3      20000110     4  103017  2184.0
4      20000110     5  102856  2169.0

## Summary:

### Data Analysis Key Findings

* The ATP ranking data from multiple decades (70s, 80s, 90s, 00s, 10s, 20s, and current) were successfully loaded and concatenated into a single DataFrame named `ranking_df`.

### Insights or Next Steps

* The combined ranking data is now ready for further analysis, such as tracking player ranking progression over time, identifying top-ranked players, or analyzing ranking trends across different eras.


# Task
Import the match results from the following URLs: "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1968.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1969.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1970.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1971.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1972.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1973.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1974.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1975.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1976.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1977.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1978.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1979.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1980.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1981.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1982.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1983.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1984.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1985.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1986.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1987.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1988.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1989.csv".

## Define urls

### Subtask:
Create a list of all the URLs for the match results data.


**Reasoning**:
Create a Python list containing the specified URLs for the match results data.



In [ ]:
match_urls = [
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1968.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1969.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1970.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1971.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1972.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1973.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1974.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1975.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1976.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1977.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1978.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1979.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1980.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1981.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1982.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1983.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1984.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1985.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1986.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1987.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1988.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1989.csv"
]

## Load and concatenate data

### Subtask:
Iterate through the list of URLs, read each CSV file into a pandas DataFrame, and concatenate all the DataFrames into a single one.


**Reasoning**:
Iterate through the list of URLs, read each CSV file into a pandas DataFrame, and concatenate all the DataFrames into a single one.



In [ ]:
import pandas as pd

match_dfs = []
for url in match_urls:
    match_dfs.append(pd.read_csv(url))

match_df = pd.concat(match_dfs, ignore_index=True)

## Display data

### Subtask:
Display the head of the combined DataFrame to verify the data is loaded correctly.


**Reasoning**:
Display the head of the combined DataFrame to verify the data is loaded correctly.



In [ ]:
display(match_df.head())

tourney_id tourney_name surface  draw_size tourney_level  tourney_date  \
0  1968-2029       Dublin   Grass       32.0             A      19680708   
1  1968-2029       Dublin   Grass       32.0             A      19680708   
2  1968-2029       Dublin   Grass       32.0             A      19680708   
3  1968-2029       Dublin   Grass       32.0             A      19680708   
4  1968-2029       Dublin   Grass       32.0             A      19680708   

   match_num  winner_id  winner_seed winner_entry  ... l_1stIn l_1stWon  \
0        270     112411          NaN          NaN  ...     NaN      NaN   
1        271     126914          NaN          NaN  ...     NaN      NaN   
2        272     209523          NaN          NaN  ...     NaN      NaN   
3        273     100084          NaN          NaN  ...     NaN      NaN   
4        274     100132          NaN          NaN  ...     NaN      NaN   

   l_2ndWon l_SvGms  l_bpSaved  l_bpFaced  winner_rank winner_rank_points  \
0       NaN     NaN        NaN        NaN          NaN                NaN   
1       NaN     NaN        NaN        NaN          NaN                NaN   
2       NaN     NaN        NaN        NaN          NaN                NaN   
3       NaN     NaN        NaN        NaN          NaN                NaN   
4       NaN     NaN        NaN        NaN          NaN                NaN   

  loser_rank loser_rank_points  
0        NaN               NaN  
1        NaN               NaN  
2        NaN               NaN  
3        NaN               NaN  
4        NaN               NaN  

[5 rows x 49 columns]

## Summary:

### Data Analysis Key Findings

*   The analysis successfully loaded and combined tennis match data from 22 separate CSV files hosted on GitHub, spanning the years 1968 to 1989, into a single pandas DataFrame named `match_df`.
*   The resulting DataFrame `match_df` contains various columns providing detailed information about each match, including tournament details (`tourney_id`, `tourney_name`, `surface`, `draw_size`, `tourney_level`, `tourney_date`), match identifier (`match_num`), and statistics for both the winner and the loser of the match.

### Insights or Next Steps

*   The consolidated DataFrame provides a comprehensive dataset for analyzing tennis match trends and player performance across over two decades.
*   The next step could involve cleaning the data, performing exploratory data analysis, or conducting statistical analysis on specific aspects of the matches or players.


# Task
Import the playerbase from "https://raw.githubusercontent.com/JeffSackmann/tennis_atp/refs/heads/master/atp_players.csv", the rankings from "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_00s.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_10s.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_20s.csv", "https://raw.githubusercontent.com/JeffSackmann/tennis_atp/refs/heads/master/atp_rankings_70s.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_80s.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_90s.csv", and "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_current.csv", and the match results from "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1968.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1969.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1970.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1971.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1972.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1973.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1974.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1975.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1976.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1977.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1978.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1979.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1980.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1981.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1982.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1983.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1984.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1985.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1986.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1987.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1988.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1989.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1990.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1991.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1992.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1993.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1994.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1995.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1996.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1997.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1998.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1999.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2000.csv", and "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2001.csv".

## Define urls

### Subtask:
Create a list of all the new URLs for the match results data.


**Reasoning**:
Create a Python list containing the specified URLs for the additional match results data (1990-2001).



In [ ]:
new_match_urls = [
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1990.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1991.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1992.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1993.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1994.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1995.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1996.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1997.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1998.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1999.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2000.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2001.csv"
]

## Load and concatenate new data

### Subtask:
Iterate through the new URLs, read each CSV file into a pandas DataFrame, and concatenate all the new DataFrames into a single one.


**Reasoning**:
Iterate through the list of new URLs, read each CSV file into a pandas DataFrame, and concatenate all the DataFrames into a single one.



In [ ]:
import pandas as pd

new_match_dfs = []
for url in new_match_urls:
    new_match_dfs.append(pd.read_csv(url))

new_match_df = pd.concat(new_match_dfs, ignore_index=True)

## Combine with existing data

### Subtask:
Concatenate the new match data DataFrame with the existing `match_df`.


**Reasoning**:
Concatenate the existing match_df with the new_match_df and assign the result back to match_df.



In [ ]:
import pandas as pd

match_df = pd.concat([match_df, new_match_df], ignore_index=True)

**Reasoning**:
Display the head of the combined DataFrame to verify the data is loaded correctly.



In [ ]:
display(match_df.head())

tourney_id tourney_name surface  draw_size tourney_level  tourney_date  \
0  1968-2029       Dublin   Grass       32.0             A      19680708   
1  1968-2029       Dublin   Grass       32.0             A      19680708   
2  1968-2029       Dublin   Grass       32.0             A      19680708   
3  1968-2029       Dublin   Grass       32.0             A      19680708   
4  1968-2029       Dublin   Grass       32.0             A      19680708   

   match_num  winner_id  winner_seed winner_entry  ... l_1stIn l_1stWon  \
0        270     112411          NaN          NaN  ...     NaN      NaN   
1        271     126914          NaN          NaN  ...     NaN      NaN   
2        272     209523          NaN          NaN  ...     NaN      NaN   
3        273     100084          NaN          NaN  ...     NaN      NaN   
4        274     100132          NaN          NaN  ...     NaN      NaN   

   l_2ndWon l_SvGms  l_bpSaved  l_bpFaced  winner_rank winner_rank_points  \
0       NaN     NaN        NaN        NaN          NaN                NaN   
1       NaN     NaN        NaN        NaN          NaN                NaN   
2       NaN     NaN        NaN        NaN          NaN                NaN   
3       NaN     NaN        NaN        NaN          NaN                NaN   
4       NaN     NaN        NaN        NaN          NaN                NaN   

  loser_rank loser_rank_points  
0        NaN               NaN  
1        NaN               NaN  
2        NaN               NaN  
3        NaN               NaN  
4        NaN               NaN  

[5 rows x 49 columns]

## Handle duplicates

### Subtask:
Check for and remove duplicate rows in the combined DataFrame.


**Reasoning**:
Check for duplicate rows in the combined DataFrame and remove them.



In [ ]:
print(f"Shape of match_df before dropping duplicates: {match_df.shape}")
duplicate_rows = match_df.duplicated().sum()
print(f"Number of duplicate rows: {duplicate_rows}")
match_df = match_df.drop_duplicates()
print(f"Shape of match_df after dropping duplicates: {match_df.shape}")

Shape of match_df before dropping duplicates: (126775, 49)
Number of duplicate rows: 0
Shape of match_df after dropping duplicates: (126775, 49)


## Display data

### Subtask:
Display the head of the combined and de-duplicated DataFrame to verify the data is loaded and cleaned correctly.


**Reasoning**:
Display the head of the combined and de-duplicated DataFrame to verify the data is loaded and cleaned correctly.



In [ ]:
display(match_df.head())

tourney_id tourney_name surface  draw_size tourney_level  tourney_date  \
0  1968-2029       Dublin   Grass       32.0             A      19680708   
1  1968-2029       Dublin   Grass       32.0             A      19680708   
2  1968-2029       Dublin   Grass       32.0             A      19680708   
3  1968-2029       Dublin   Grass       32.0             A      19680708   
4  1968-2029       Dublin   Grass       32.0             A      19680708   

   match_num  winner_id  winner_seed winner_entry  ... l_1stIn l_1stWon  \
0        270     112411          NaN          NaN  ...     NaN      NaN   
1        271     126914          NaN          NaN  ...     NaN      NaN   
2        272     209523          NaN          NaN  ...     NaN      NaN   
3        273     100084          NaN          NaN  ...     NaN      NaN   
4        274     100132          NaN          NaN  ...     NaN      NaN   

   l_2ndWon l_SvGms  l_bpSaved  l_bpFaced  winner_rank winner_rank_points  \
0       NaN     NaN        NaN        NaN          NaN                NaN   
1       NaN     NaN        NaN        NaN          NaN                NaN   
2       NaN     NaN        NaN        NaN          NaN                NaN   
3       NaN     NaN        NaN        NaN          NaN                NaN   
4       NaN     NaN        NaN        NaN          NaN                NaN   

  loser_rank loser_rank_points  
0        NaN               NaN  
1        NaN               NaN  
2        NaN               NaN  
3        NaN               NaN  
4        NaN               NaN  

[5 rows x 49 columns]

## Summary:

### Data Analysis Key Findings

*   The match results data from 1990 to 2001 were successfully loaded and concatenated into a single DataFrame (`new_match_df`).
*   The newly loaded match data (`new_match_df`) was successfully combined with the existing match data (`match_df`).
*   No duplicate rows were found in the combined match data.

### Insights or Next Steps

*   The combined match data from 1968 to 2001 is now ready for comprehensive analysis, such as identifying trends in match outcomes over time or analyzing player performance across different eras.
*   Further data cleaning or feature engineering could be performed on the combined match data to prepare it for more advanced modeling or analysis tasks.


# Task
Import the playerbase from "https://raw.githubusercontent.com/JeffSackmann/tennis_atp/refs/heads/master/atp_players.csv", the rankings from "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_00s.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_10s.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_20s.csv", "https://raw.githubusercontent.com/JeffSackmann/tennis_atp/refs/heads/master/atp_rankings_70s.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_80s.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_90s.csv", and "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_current.csv", and the match results from "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1968.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1969.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1970.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1971.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1972.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1973.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1974.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1975.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1976.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1977.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1978.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1979.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1980.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1981.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1982.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1983.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1984.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1985.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1986.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1987.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1988.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1989.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1990.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1991.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1992.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1993.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1994.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1995.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1996.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1997.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1998.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1999.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2000.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2001.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2002.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2003.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2004.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2005.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2006.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2007.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2008.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2009.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2010.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2011.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2012.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2013.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2014.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2015.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2016.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2017.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2018.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2019.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2020.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2021.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2022.csv", and "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2023.csv".

## Define urls

### Subtask:
Create a list of all the new URLs for the match results data (2002-2023).


**Reasoning**:
Create a Python list containing the specified URLs for the additional match results data (2002-2023).



In [ ]:
newer_match_urls = [
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2002.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2003.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2004.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2005.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2006.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2007.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2008.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2009.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2010.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2011.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2012.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2013.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2014.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2015.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2016.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2017.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2018.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2019.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2020.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2021.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2022.csv",
    "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2023.csv"
]

## Prepare Data

### Subtask:
Select relevant features and target variable from `match_df` and split the data into training and testing sets (60% training, 40% testing).

**Reasoning**:
Select relevant columns from `match_df` for training.
Define the target variable as the winner of the match.
Split the data into training and testing sets using a 60/40 split.

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split

# Select features and target variable
features = ['winner_rank', 'loser_rank', 'winner_rank_points', 'loser_rank_points', 'surface'] # Corrected column names
target = 'winner_id' # Or you could predict the outcome (winner/loser) directly

# Drop rows with missing values in the selected columns for simplicity
match_df_cleaned = match_df.dropna(subset=features + [target])

# Convert categorical features to numerical using one-hot encoding
match_df_cleaned = pd.get_dummies(match_df_cleaned, columns=['surface'], dummy_na=False)

# Update features list with the new one-hot encoded columns
features = [col for col in match_df_cleaned.columns if col.startswith(('winner_rank', 'loser_rank', 'winner_rank_points', 'loser_rank_points', 'surface_'))]


X = match_df_cleaned[features]
y = match_df_cleaned[target]

# Split data into training and testing sets (60% training)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=42)

print(f"Shape of X_train: {X_train.shape}")
print(f"Shape of X_test: {X_test.shape}")
print(f"Shape of y_train: {y_train.shape}")
print(f"Shape of y_test: {y_test.shape}")

Shape of X_train: (25722, 8)
Shape of X_test: (17149, 8)
Shape of y_train: (25722,)
Shape of y_test: (17149,)


In [ ]:
!pip install xgboost

## Load and concatenate new data

### Subtask:
Iterate through the new URLs, read each CSV file into a pandas DataFrame, and concatenate all the new DataFrames into a single one.


**Reasoning**:
Iterate through the new URLs, read each CSV file into a pandas DataFrame, and concatenate all the DataFrames into a single one.



In [ ]:
newer_match_dfs = []
for url in newer_match_urls:
    newer_match_dfs.append(pd.read_csv(url))

newer_match_df = pd.concat(newer_match_dfs, ignore_index=True)

## Combine with existing data

### Subtask:
Concatenate the new match data DataFrame with the existing `match_df`.


**Reasoning**:
Concatenate the existing match_df with the newer_match_df and assign the result back to match_df.



In [ ]:
import pandas as pd

match_df = pd.concat([match_df, newer_match_df], ignore_index=True)

**Reasoning**:
Display the head of the combined DataFrame to verify the data is loaded correctly.



In [ ]:
display(match_df.head())

tourney_id tourney_name surface  draw_size tourney_level  tourney_date  \
0  1968-2029       Dublin   Grass       32.0             A      19680708   
1  1968-2029       Dublin   Grass       32.0             A      19680708   
2  1968-2029       Dublin   Grass       32.0             A      19680708   
3  1968-2029       Dublin   Grass       32.0             A      19680708   
4  1968-2029       Dublin   Grass       32.0             A      19680708   

   match_num  winner_id  winner_seed winner_entry  ... l_1stIn l_1stWon  \
0        270     112411          NaN          NaN  ...     NaN      NaN   
1        271     126914          NaN          NaN  ...     NaN      NaN   
2        272     209523          NaN          NaN  ...     NaN      NaN   
3        273     100084          NaN          NaN  ...     NaN      NaN   
4        274     100132          NaN          NaN  ...     NaN      NaN   

   l_2ndWon l_SvGms  l_bpSaved  l_bpFaced  winner_rank winner_rank_points  \
0       NaN     NaN        NaN        NaN          NaN                NaN   
1       NaN     NaN        NaN        NaN          NaN                NaN   
2       NaN     NaN        NaN        NaN          NaN                NaN   
3       NaN     NaN        NaN        NaN          NaN                NaN   
4       NaN     NaN        NaN        NaN          NaN                NaN   

  loser_rank loser_rank_points  
0        NaN               NaN  
1        NaN               NaN  
2        NaN               NaN  
3        NaN               NaN  
4        NaN               NaN  

[5 rows x 49 columns]

**Reasoning**:
Check for and remove duplicate rows in the combined DataFrame.



In [ ]:
print(f"Shape of match_df before dropping duplicates: {match_df.shape}")
duplicate_rows = match_df.duplicated().sum()
print(f"Number of duplicate rows: {duplicate_rows}")
match_df = match_df.drop_duplicates()
print(f"Shape of match_df after dropping duplicates: {match_df.shape}")

Shape of match_df before dropping duplicates: (191920, 49)
Number of duplicate rows: 0
Shape of match_df after dropping duplicates: (191920, 49)


**Reasoning**:
Display the head of the combined and de-duplicated DataFrame to verify the data is loaded and cleaned correctly.



In [ ]:
display(match_df.head())

tourney_id tourney_name surface  draw_size tourney_level  tourney_date  \
0  1968-2029       Dublin   Grass       32.0             A      19680708   
1  1968-2029       Dublin   Grass       32.0             A      19680708   
2  1968-2029       Dublin   Grass       32.0             A      19680708   
3  1968-2029       Dublin   Grass       32.0             A      19680708   
4  1968-2029       Dublin   Grass       32.0             A      19680708   

   match_num  winner_id  winner_seed winner_entry  ... l_1stIn l_1stWon  \
0        270     112411          NaN          NaN  ...     NaN      NaN   
1        271     126914          NaN          NaN  ...     NaN      NaN   
2        272     209523          NaN          NaN  ...     NaN      NaN   
3        273     100084          NaN          NaN  ...     NaN      NaN   
4        274     100132          NaN          NaN  ...     NaN      NaN   

   l_2ndWon l_SvGms  l_bpSaved  l_bpFaced  winner_rank winner_rank_points  \
0       NaN     NaN        NaN        NaN          NaN                NaN   
1       NaN     NaN        NaN        NaN          NaN                NaN   
2       NaN     NaN        NaN        NaN          NaN                NaN   
3       NaN     NaN        NaN        NaN          NaN                NaN   
4       NaN     NaN        NaN        NaN          NaN                NaN   

  loser_rank loser_rank_points  
0        NaN               NaN  
1        NaN               NaN  
2        NaN               NaN  
3        NaN               NaN  
4        NaN               NaN  

[5 rows x 49 columns]

## Summary:

### Data Analysis Key Findings

*   The analysis successfully imported and combined tennis player, ranking, and match data from various CSV files hosted on GitHub, spanning multiple decades up to 2023.
*   The combined match results DataFrame (`match_df`) contains 191,920 rows and 49 columns after concatenating data from 1968 through 2023.
*   No duplicate rows were found in the final combined `match_df` DataFrame.

### Insights or Next Steps

*   The comprehensive match dataset is now ready for further analysis, such as calculating player win-loss records, analyzing tournament performance over time, or exploring trends in match statistics.
*   Further steps could involve merging the match data with the player and ranking data to enable richer analysis that incorporates player attributes and historical rankings.


# Task
Import the playerbase from "https://raw.githubusercontent.com/JeffSackmann/tennis_atp/refs/heads/master/atp_players.csv", the rankings from "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_00s.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_10s.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_20s.csv", "https://raw.githubusercontent.com/JeffSackmann/tennis_atp/refs/heads/master/atp_rankings_70s.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_80s.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_90s.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_rankings_current.csv", and the match results from "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1968.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1969.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1970.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1971.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1972.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1973.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1974.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1975.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1976.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1977.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1978.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1979.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1980.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1981.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1982.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1983.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1984.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1985.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1986.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1987.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1988.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1989.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1990.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1991.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1992.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1993.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1994.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1995.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1996.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1997.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1998.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1999.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2000.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1980.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1981.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1982.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1983.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1984.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1985.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1986.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1987.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1988.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_1989.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2001.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2002.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2003.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2004.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2005.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2006.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2007.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2008.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2009.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2010.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2011.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2012.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2013.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2014.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2015.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2016.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2017.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2018.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2019.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2020.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2021.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2022.csv", "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2023.csv", and "https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2024.csv".

## Define urls

### Subtask:
Create a list with the URL for the 2024 match results data.


**Reasoning**:
Create a Python list containing the specified URL for the 2024 match results data.



In [ ]:
match_urls_2024 = ["https://github.com/JeffSackmann/tennis_atp/raw/refs/heads/master/atp_matches_2024.csv"]

## Load new data

### Subtask:
Read the CSV file from the URL into a pandas DataFrame.


**Reasoning**:
Read the CSV file from the first URL in the `match_urls_2024` list into a pandas DataFrame and display the head.



In [ ]:
match_df_2024 = pd.read_csv(match_urls_2024[0])
display(match_df_2024.head())

tourney_id tourney_name surface  draw_size tourney_level  tourney_date  \
0  2024-0339     Brisbane    Hard         32             A      20240101   
1  2024-0339     Brisbane    Hard         32             A      20240101   
2  2024-0339     Brisbane    Hard         32             A      20240101   
3  2024-0339     Brisbane    Hard         32             A      20240101   
4  2024-0339     Brisbane    Hard         32             A      20240101   

   match_num  winner_id  winner_seed winner_entry  ... l_1stIn l_1stWon  \
0        300     105777          2.0          NaN  ...    58.0     44.0   
1        299     208029          1.0          NaN  ...    35.0     31.0   
2        298     105777          2.0          NaN  ...    39.0     24.0   
3        297     208029          1.0          NaN  ...    51.0     31.0   
4        296     126128          NaN          NaN  ...    37.0     27.0   

   l_2ndWon l_SvGms  l_bpSaved  l_bpFaced  winner_rank winner_rank_points  \
0      16.0    11.0        8.0        9.0         14.0             2570.0   
1      10.0    11.0        5.0        7.0          8.0             3660.0   
2      14.0    10.0        5.0        7.0         14.0             2570.0   
3      16.0    10.0        3.0        5.0          8.0             3660.0   
4      16.0    10.0        5.0        8.0         39.0             1122.0   

  loser_rank loser_rank_points  
0        8.0            3660.0  
1       39.0            1122.0  
2       55.0             902.0  
3      116.0             573.0  
4       44.0            1021.0  

[5 rows x 49 columns]

## Combine with existing data

### Subtask:
Concatenate the new match data DataFrame with the existing `match_df`.


**Reasoning**:
Concatenate the existing match_df with the match_df_2024 and assign the result back to match_df.



In [ ]:
import pandas as pd

match_df = pd.concat([match_df, match_df_2024], ignore_index=True)

**Reasoning**:
Display the head of the combined DataFrame to verify the data is loaded correctly.



In [ ]:
display(match_df.head())

tourney_id tourney_name surface  draw_size tourney_level  tourney_date  \
0  1968-2029       Dublin   Grass       32.0             A      19680708   
1  1968-2029       Dublin   Grass       32.0             A      19680708   
2  1968-2029       Dublin   Grass       32.0             A      19680708   
3  1968-2029       Dublin   Grass       32.0             A      19680708   
4  1968-2029       Dublin   Grass       32.0             A      19680708   

   match_num  winner_id  winner_seed winner_entry  ... l_1stIn l_1stWon  \
0        270     112411          NaN          NaN  ...     NaN      NaN   
1        271     126914          NaN          NaN  ...     NaN      NaN   
2        272     209523          NaN          NaN  ...     NaN      NaN   
3        273     100084          NaN          NaN  ...     NaN      NaN   
4        274     100132          NaN          NaN  ...     NaN      NaN   

   l_2ndWon l_SvGms  l_bpSaved  l_bpFaced  winner_rank winner_rank_points  \
0       NaN     NaN        NaN        NaN          NaN                NaN   
1       NaN     NaN        NaN        NaN          NaN                NaN   
2       NaN     NaN        NaN        NaN          NaN                NaN   
3       NaN     NaN        NaN        NaN          NaN                NaN   
4       NaN     NaN        NaN        NaN          NaN                NaN   

  loser_rank loser_rank_points  
0        NaN               NaN  
1        NaN               NaN  
2        NaN               NaN  
3        NaN               NaN  
4        NaN               NaN  

[5 rows x 49 columns]

**Reasoning**:
Check for and remove duplicate rows in the combined DataFrame.



In [ ]:
print(f"Shape of match_df before dropping duplicates: {match_df.shape}")
duplicate_rows = match_df.duplicated().sum()
print(f"Number of duplicate rows: {duplicate_rows}")
match_df = match_df.drop_duplicates()
print(f"Shape of match_df after dropping duplicates: {match_df.shape}")

Shape of match_df before dropping duplicates: (194996, 49)
Number of duplicate rows: 0
Shape of match_df after dropping duplicates: (194996, 49)


**Reasoning**:
Display the head of the combined and de-duplicated DataFrame to verify the data is loaded and cleaned correctly.



In [ ]:
display(match_df.head())

tourney_id tourney_name surface  draw_size tourney_level  tourney_date  \
0  1968-2029       Dublin   Grass       32.0             A      19680708   
1  1968-2029       Dublin   Grass       32.0             A      19680708   
2  1968-2029       Dublin   Grass       32.0             A      19680708   
3  1968-2029       Dublin   Grass       32.0             A      19680708   
4  1968-2029       Dublin   Grass       32.0             A      19680708   

   match_num  winner_id  winner_seed winner_entry  ... l_1stIn l_1stWon  \
0        270     112411          NaN          NaN  ...     NaN      NaN   
1        271     126914          NaN          NaN  ...     NaN      NaN   
2        272     209523          NaN          NaN  ...     NaN      NaN   
3        273     100084          NaN          NaN  ...     NaN      NaN   
4        274     100132          NaN          NaN  ...     NaN      NaN   

   l_2ndWon l_SvGms  l_bpSaved  l_bpFaced  winner_rank winner_rank_points  \
0       NaN     NaN        NaN        NaN          NaN                NaN   
1       NaN     NaN        NaN        NaN          NaN                NaN   
2       NaN     NaN        NaN        NaN          NaN                NaN   
3       NaN     NaN        NaN        NaN          NaN                NaN   
4       NaN     NaN        NaN        NaN          NaN                NaN   

  loser_rank loser_rank_points  
0        NaN               NaN  
1        NaN               NaN  
2        NaN               NaN  
3        NaN               NaN  
4        NaN               NaN  

[5 rows x 49 columns]

## Summary:

### Data Analysis Key Findings

* The 2024 match data from the provided URL was successfully loaded into a pandas DataFrame.
* The 2024 match data was successfully concatenated with the existing match data, resulting in a combined DataFrame.
* No duplicate rows were found in the combined match data after concatenation.

### Insights or Next Steps

* The combined dataset containing match results from 1968 to 2024 is now ready for further analysis, such as examining player performance trends over time or analyzing match statistics.
* Potential next steps could include exploring specific aspects of the 2024 season, like identifying top-performing players or analyzing tournament outcomes.
